staging/src/k8s.io/apiserver/pkg/util/openapi/enablement.go

```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"strings"
+
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/kube-openapi/pkg/common"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+// enumTypeDescriptionHeader is the header of enum section in schema description.
+const enumTypeDescriptionHeader = "Possible enum values:"
+
+// WrapGetOpenAPIDefinitions wraps a GetOpenAPIDefinitions to revert
+// any change to the schema that was made by a disabled feature.
+func WrapGetOpenAPIDefinitions(GetOpenAPIDefinitions common.GetOpenAPIDefinitions) common.GetOpenAPIDefinitions {
+	return func(ref common.ReferenceCallback) map[string]common.OpenAPIDefinition {
+		defs := GetOpenAPIDefinitions(ref)
+		amendDefinitions(defs)
+		return defs
+	}
+}
+
+func amendDefinitions(defs map[string]common.OpenAPIDefinition) {
+	// revert changes from OpenAPIEnums
+	if !utilfeature.DefaultFeatureGate.Enabled(genericfeatures.OpenAPIEnums) {
+		for gvk, def := range defs {
+			pruneEnums(&def.Schema)
+			defs[gvk] = def
+		}
+	}
+}
+
+func pruneEnums(schema *spec.Schema) {
+	if headerIndex := strings.Index(schema.Description, enumTypeDescriptionHeader); headerIndex != -1 {
+		// remove the enum section from description.
+		// note that the new lines before the header should be removed too,
+		// thus the slice range.
+		schema.Description = schema.Description[:headerIndex]
+	}
+	schema.Enum = nil
+	for name, prop := range schema.Properties {
```

Comment:
Are we sure we only need to descend into properties? Or are there schemas that can be nested outside properties? What about `additionalProperties` for maps? And what about lists?

Reply:
If you run Joe's test mentioned somewhere else, you should be able to find a few of these maybe.

Reply:
To be fair, I know we might not have schemas in all the places we have schemas today, but for the sake of being future-proof, I'd just be exhaustive and recurse into all the places that can have a schema, e.g. https://github.com/kubernetes/kube-openapi/blob/master/pkg/handler/default_pruning.go#L42-L208 (that code is a lot more convoluted because we can't change in place and we're trying to avoid copies, but that gives you an idea of all the places where we can have a schema hidden).

Reply:
Thanks! I think iterating through all schemas should be the right choice.

As a side note. Just did a quick go through all our built-ins. Here is all JSON-path to an enum field.

```python3
#!/usr/bin/env python3

import json

def dfs(prefix, obj):
    if isinstance(obj, dict):
        if 'enum' in obj:
            print(prefix)
        for k, v in obj.items():
            dfs('{}/{}'.format(prefix, k), v)
    if isinstance(obj, list):
        for v in obj:
            dfs(prefix + '[]/', v)

if __name__ == '__main__':
    with open('./api/openapi-spec/swagger.json') as input:
        defs = json.load(input)['definitions']
        dfs('.', defs)

```
```text
./io.k8s.api.apiserverinternal.v1alpha1.StorageVersionCondition/properties/status
./io.k8s.api.apiserverinternal.v1alpha1.StorageVersionCondition/properties/type
./io.k8s.api.apps.v1.DaemonSetCondition/properties/status
./io.k8s.api.apps.v1.DaemonSetUpdateStrategy/properties/type
./io.k8s.api.apps.v1.DeploymentCondition/properties/status
./io.k8s.api.apps.v1.DeploymentCondition/properties/type
./io.k8s.api.apps.v1.DeploymentStrategy/properties/type
./io.k8s.api.apps.v1.ReplicaSetCondition/properties/status
./io.k8s.api.apps.v1.ReplicaSetCondition/properties/type
./io.k8s.api.apps.v1.StatefulSetCondition/properties/status
./io.k8s.api.apps.v1.StatefulSetSpec/properties/podManagementPolicy
./io.k8s.api.apps.v1.StatefulSetUpdateStrategy/properties/type
./io.k8s.api.autoscaling.v2beta1.ContainerResourceMetricSource/properties/name
./io.k8s.api.autoscaling.v2beta1.ContainerResourceMetricStatus/properties/name
./io.k8s.api.autoscaling.v2beta1.HorizontalPodAutoscalerCondition/properties/status
./io.k8s.api.autoscaling.v2beta1.HorizontalPodAutoscalerCondition/properties/type
./io.k8s.api.autoscaling.v2beta1.MetricSpec/properties/type
./io.k8s.api.autoscaling.v2beta1.MetricStatus/properties/type
./io.k8s.api.autoscaling.v2beta1.ResourceMetricSource/properties/name
./io.k8s.api.autoscaling.v2beta1.ResourceMetricStatus/properties/name
./io.k8s.api.autoscaling.v2beta2.ContainerResourceMetricSource/properties/name
./io.k8s.api.autoscaling.v2beta2.ContainerResourceMetricStatus/properties/name
./io.k8s.api.autoscaling.v2beta2.HPAScalingPolicy/properties/type
./io.k8s.api.autoscaling.v2beta2.HorizontalPodAutoscalerCondition/properties/status
./io.k8s.api.autoscaling.v2beta2.HorizontalPodAutoscalerCondition/properties/type
./io.k8s.api.autoscaling.v2beta2.MetricSpec/properties/type
./io.k8s.api.autoscaling.v2beta2.MetricStatus/properties/type
./io.k8s.api.autoscaling.v2beta2.MetricTarget/properties/type
./io.k8s.api.autoscaling.v2beta2.ResourceMetricSource/properties/name
./io.k8s.api.autoscaling.v2beta2.ResourceMetricStatus/properties/name
./io.k8s.api.batch.v1.CronJobSpec/properties/concurrencyPolicy
./io.k8s.api.batch.v1.JobCondition/properties/status
./io.k8s.api.batch.v1.JobCondition/properties/type
./io.k8s.api.batch.v1beta1.CronJobSpec/properties/concurrencyPolicy
./io.k8s.api.certificates.v1.CertificateSigningRequestCondition/properties/status
./io.k8s.api.certificates.v1.CertificateSigningRequestCondition/properties/type
./io.k8s.api.core.v1.ComponentCondition/properties/status
./io.k8s.api.core.v1.ComponentCondition/properties/type
./io.k8s.api.core.v1.Container/properties/imagePullPolicy
./io.k8s.api.core.v1.Container/properties/terminationMessagePolicy
./io.k8s.api.core.v1.ContainerPort/properties/protocol
./io.k8s.api.core.v1.EndpointPort/properties/protocol
./io.k8s.api.core.v1.EphemeralContainer/properties/imagePullPolicy
./io.k8s.api.core.v1.EphemeralContainer/properties/terminationMessagePolicy
./io.k8s.api.core.v1.HTTPGetAction/properties/scheme
./io.k8s.api.core.v1.LimitRangeItem/properties/type
./io.k8s.api.core.v1.NamespaceCondition/properties/status
./io.k8s.api.core.v1.NamespaceCondition/properties/type
./io.k8s.api.core.v1.NamespaceStatus/properties/phase
./io.k8s.api.core.v1.NodeAddress/properties/type
./io.k8s.api.core.v1.NodeCondition/properties/status
./io.k8s.api.core.v1.NodeCondition/properties/type
./io.k8s.api.core.v1.NodeSelectorRequirement/properties/operator
./io.k8s.api.core.v1.NodeStatus/properties/phase
./io.k8s.api.core.v1.PersistentVolumeClaimCondition/properties/status
./io.k8s.api.core.v1.PersistentVolumeClaimCondition/properties/type
./io.k8s.api.core.v1.PersistentVolumeClaimStatus/properties/phase
./io.k8s.api.core.v1.PersistentVolumeSpec/properties/persistentVolumeReclaimPolicy
./io.k8s.api.core.v1.PersistentVolumeStatus/properties/phase
./io.k8s.api.core.v1.PodCondition/properties/status
./io.k8s.api.core.v1.PodCondition/properties/type
./io.k8s.api.core.v1.PodReadinessGate/properties/conditionType
./io.k8s.api.core.v1.PodSpec/properties/dnsPolicy
./io.k8s.api.core.v1.PodSpec/properties/restartPolicy
./io.k8s.api.core.v1.PodStatus/properties/phase
./io.k8s.api.core.v1.PodStatus/properties/qosClass
./io.k8s.api.core.v1.PortStatus/properties/protocol
./io.k8s.api.core.v1.ReplicationControllerCondition/properties/status
./io.k8s.api.core.v1.ReplicationControllerCondition/properties/type
./io.k8s.api.core.v1.ScopedResourceSelectorRequirement/properties/operator
./io.k8s.api.core.v1.ScopedResourceSelectorRequirement/properties/scopeName
./io.k8s.api.core.v1.SeccompProfile/properties/type
./io.k8s.api.core.v1.Secret/properties/type
./io.k8s.api.core.v1.ServicePort/properties/protocol
./io.k8s.api.core.v1.ServiceSpec/properties/externalTrafficPolicy
./io.k8s.api.core.v1.ServiceSpec/properties/sessionAffinity
./io.k8s.api.core.v1.ServiceSpec/properties/type
./io.k8s.api.core.v1.Taint/properties/effect
./io.k8s.api.core.v1.Toleration/properties/effect
./io.k8s.api.core.v1.Toleration/properties/operator
./io.k8s.api.core.v1.TopologySpreadConstraint/properties/whenUnsatisfiable
./io.k8s.api.discovery.v1.EndpointSlice/properties/addressType
./io.k8s.api.discovery.v1beta1.EndpointSlice/properties/addressType
./io.k8s.api.flowcontrol.v1beta1.FlowDistinguisherMethod/properties/type
./io.k8s.api.flowcontrol.v1beta1.FlowSchemaCondition/properties/status
./io.k8s.api.flowcontrol.v1beta1.LimitResponse/properties/type
./io.k8s.api.flowcontrol.v1beta1.PriorityLevelConfigurationCondition/properties/status
./io.k8s.api.flowcontrol.v1beta1.PriorityLevelConfigurationSpec/properties/type
./io.k8s.api.flowcontrol.v1beta1.Subject/properties/kind
./io.k8s.api.flowcontrol.v1beta2.FlowDistinguisherMethod/properties/type
./io.k8s.api.flowcontrol.v1beta2.FlowSchemaCondition/properties/status
./io.k8s.api.flowcontrol.v1beta2.LimitResponse/properties/type
./io.k8s.api.flowcontrol.v1beta2.PriorityLevelConfigurationCondition/properties/status
./io.k8s.api.flowcontrol.v1beta2.PriorityLevelConfigurationSpec/properties/type
./io.k8s.api.flowcontrol.v1beta2.Subject/properties/kind
./io.k8s.api.policy.v1beta1.FSGroupStrategyOptions/properties/rule
./io.k8s.api.policy.v1beta1.RunAsGroupStrategyOptions/properties/rule
./io.k8s.api.policy.v1beta1.RunAsUserStrategyOptions/properties/rule
./io.k8s.api.policy.v1beta1.SELinuxStrategyOptions/properties/rule
./io.k8s.api.policy.v1beta1.SupplementalGroupsStrategyOptions/properties/rule
./io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.CustomResourceConversion/properties/strategy
./io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.CustomResourceDefinitionCondition/properties/status
./io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.CustomResourceDefinitionCondition/properties/type
./io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.CustomResourceDefinitionSpec/properties/scope
./io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.JSONSchemaProps/properties
./io.k8s.apimachinery.pkg.apis.meta.v1.LabelSelectorRequirement/properties/operator
./io.k8s.apimachinery.pkg.apis.meta.v1.ManagedFieldsEntry/properties/operation
./io.k8s.apimachinery.pkg.apis.meta.v1.Status/properties/reason
./io.k8s.apimachinery.pkg.apis.meta.v1.StatusCause/properties/reason
./io.k8s.kube-aggregator.pkg.apis.apiregistration.v1.APIServiceCondition/properties/status
./io.k8s.kube-aggregator.pkg.apis.apiregistration.v1.APIServiceCondition/properties/type

```

Reply:
Yeah, thanks. From a cursory look, it looks like they all come from properties but going through all the paths will be better for whoever starts using enum somewhere else. Thanks!